### ISSUE TITLE
Deserialization of generic classes with different type arguments

### ISSUE THREAD

**Issue by user85473** [FasterXML/jackson-databind#2068]
I was writing cool generic json configs when I realized that jackson deserialize class regardless of type arguments boundaries.
Consider this:
```
    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
    @JsonSubTypes({
            @JsonSubTypes.Type(value = AInteger.class, name = "AInteger"),
            @JsonSubTypes.Type(value = AString.class, name = "AString") })
    private static class A<T> {
        @JsonProperty
        private T t;
    }

    private static class AInteger extends A<Integer> {
    }

    private static class AString extends A<String> {
    }
```
Then I try to deserialize this json:
`{"type":"AInteger", "t": 123}`
like:
```
A<String> o = new ObjectMapper().readValue(json, new TypeReference<A<String>>() { });
```

AInteger is not a subclass of A\<String\>, and we have this information in runtime, but deserialization succeed.
It's really disappointing cause even if deserialization of some complex config succeed it still could be semantically incorrect.

Tested on Jackson version: 2.9.6

**Comment by cowtowncoder** [comment]
Thank you for reporting the problem with all the necessary information.
I'll have to see whether this combination can be supported: generics and polymorphic type information are unfortunately tricky combination.

**Comment by user85473** [comment]
Thanks!

Hope it will help. I noticed strange behaviour while debugging jackson code.
If we change the example above a little bit:
```
    // just adding useless type parameter
    private static class AInteger<T> extends A<Integer> {
    }

    // same here for research purposes
    private static class AString<T> extends A<String> {
    }
```
everything starts to work as expected.
I get an exception when deserializing AInteger as A\<String\> and everything goes smooth when deserializing AString as A\<String\>.

It's because of these lines of constructSpecializedType method from TypeFactory:
```
            // (3) Sub-class does not take type parameters -- just resolve subtype
            int typeParamCount = subclass.getTypeParameters().length;
            if (typeParamCount == 0) {
                newType = _fromClass(null, subclass, EMPTY_BINDINGS);
                break;
            }
```

**Comment by cowtowncoder** [comment]
A few things have been fixed as of 2.12.0 in context of type variable resolution.
At this point would need a full reproduction against 2.12.0 to show the issue -- may be reopened with one.
